File: Sources/Providers/ExeProvider.swift
import Foundation

/// exe.dev as a `VMProvider`: the HTTPS `/exec` API for VM lifecycle and GitHub
/// integrations, the `llm.int.exe.xyz` gateway, SSH against `<name>.exe.xyz`,
/// and the reflection endpoint that lets a VM name itself.
final class ExeProvider: VMProvider {
    let id: VMProviderID = .exe
    let displayName = "exe.dev"
    let defaultBrowserURL = "https://exe.dev"
    let supportsAutoNaming = true
    let tokenEnvVar = "EXE_DEV_TOKEN"

    private let service: ExeService
    private let tokenProvider: () -> String

    init(tokenProvider: @escaping () -> String = { "" }) {
        self.tokenProvider = tokenProvider
        service = ExeService(client: ExeClient(tokenProvider: tokenProvider))
    }

    func effectiveToken() -> String { tokenProvider() }

    // MARK: - LLM gateway

    func listModels() async -> (models: [GatewayModel], error: String?) {
        await LLMGateway.list(config: .exe)
    }

    func harnessWiring(for model: GatewayModel) -> HarnessWiring {
        LLMGateway.wiring(for: model, config: .exe)
    }

    // MARK: - VM lifecycle

    func listVMs() async throws -> [RemoteVMRecord] {
        try await service.listVMs().map(Self.record(from:))
    }

    func createVM(name: String, tags: [String], environment: [EnvVar]) async throws -> RemoteVMRecord {
        let vm = try await service.createVM(name: name, tags: tags, environment: environment)
        return Self.record(from: vm)
    }

    func deleteVM(name: String) async throws {
        try await service.deleteVM(name: name)
    }

    // MARK: - GitHub

    /// Ensure a GitHub integration exists for each repo (creating one that acts
    /// as the user if not) and return the per-repo tags that bind them to the
    /// new VM. exe.dev brokers clone access through those integrations and the
    /// `github.int.exe.xyz` proxy, so no clone credentials go in the environment.
    func prepareGitHub(repos: [String]) async throws -> GitHubSetup {
        guard !repos.isEmpty else {
            return GitHubSetup(tags: [], cloneEnvironment: [], clone: .exe)
        }
        let existing = try await service.listIntegrations()
        var tags: [String] = []
        for repo in repos {
            tags.append(try await service.ensureGithubIntegration(repo: repo, existing: existing))
        }
        return GitHubSetup(tags: tags, cloneEnvironment: [], clone: .exe)
    }

    // MARK: - Auto-naming

    func generateRenameToken() async throws -> String {
        try await service.generateRenameToken()
    }

    // MARK: - Naming and reachability

    func destination(forVMName name: String) -> String { "\(name).exe.xyz" }

    func webURL(forDestination destination: String) -> String? { "https://\(destination)" }

    /// Shelley — exe.dev's own web agent — served on the VM's port 9999.
    func shelleyURL(forDestination destination: String) -> String? { "https://\(destination):9999/" }

    /// The reflection integration — attached to every VM on an account by
    /// default — publishes the VM's current name at the top level of its index.
    let reflectionNameCommand: String? = "curl -fsS --max-time 5 https://reflection.int.exe.xyz/"

    func parseReflectedName(_ output: String) -> String? {
        guard let data = output.data(using: .utf8),
              let object = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let name = object["name"] as? String,
              !name.isEmpty
        else { return nil }
        return name
    }

    // MARK: - Transport

    func transport(forDestination destination: String) -> RemoteTransport {
        SSHTransport(destination: destination)
    }

    // MARK: - Mapping

    /// An `ExeVM` as the app-agnostic `RemoteVMRecord`. `ssh_dest` is the
    /// transport handle; the web URL is the SSH host over HTTPS, matching what
    /// `TerminalSession.webURL` always produced.
    static func record(from vm: ExeVM) -> RemoteVMRecord {
        let name = vm.vm_name ?? vm.ssh_dest ?? "unknown"
        let destination = vm.ssh_dest ?? (vm.vm_name.map { "\($0).exe.xyz" } ?? "unknown")
        return RemoteVMRecord(
            name: name,
            destination: destination,
            webURL: "https://\(destination)",
            status: vm.status)
    }
}

File: src/providers/exe-provider.ts
/**
 * exe.dev as a `VMProvider`: the HTTPS `/exec` API for VM lifecycle and GitHub
 * integrations, the `llm.int.exe.xyz` gateway, SSH against `<name>.exe.xyz`,
 * and the reflection endpoint that lets a VM name itself.
 */

import { ExeClient } from "./exe-client.ts";
import { ExeService, type ExeVM } from "./exe-service.ts";
import { SSHTransport } from "./ssh-transport.ts";
import { tokenGitHubSetup } from "./github-setup.ts";
import { type GatewayModel, gatewayWiring, listGatewayModels } from "../model/llm-gateway.ts";
import {
  EXE_GATEWAY,
  type GitHubSetup,
  type HarnessWiring,
  type ModelListing,
  type ProviderCredential,
  type RemoteTransport,
  type RemoteVMRecord,
  type VMProvider,
  type VMProviderID,
} from "./types.ts";
import type { EnvVar } from "../config/env-var.ts";

export class ExeProvider implements VMProvider {
  readonly id: VMProviderID = "exe";
  readonly displayName = "exe.dev";
  readonly defaultBrowserURL = "https://exe.dev";
  readonly supportsAutoNaming = true;
  readonly credential: ProviderCredential = { kind: "token", envVar: "EXE_DEV_TOKEN" };

  /**
   * The reflection integration — attached to every VM on an account by default
   * — publishes the VM's current name at the top level of its index.
   */
  readonly reflectionNameCommand = "curl -fsS --max-time 5 https://reflection.int.exe.xyz/";

  private service: ExeService;

  constructor(private tokenProvider: () => string) {
    this.service = new ExeService(new ExeClient(tokenProvider));
  }

  effectiveToken(): string {
    return this.tokenProvider();
  }

  checkAvailable(): Promise<boolean> {
    return Promise.resolve(this.effectiveToken() !== "");
  }

  // MARK: - LLM gateway

  listModels(): Promise<ModelListing> {
    return listGatewayModels(EXE_GATEWAY);
  }

  harnessWiring(model: GatewayModel): HarnessWiring {
    return gatewayWiring(model, EXE_GATEWAY);
  }

  // MARK: - VM lifecycle

  async listVMs(): Promise<RemoteVMRecord[]> {
    return (await this.service.listVMs()).map(recordFromExeVM);
  }

  async createVM(name: string, tags: string[], environment: EnvVar[]): Promise<RemoteVMRecord> {
    return recordFromExeVM(await this.service.createVM(name, tags, environment));
  }

  deleteVM(name: string): Promise<void> {
    return this.service.deleteVM(name);
  }

  // MARK: - GitHub

  /**
   * The same token every other provider clones with, rather than exe.dev's own
   * GitHub integrations: one setup path, and nothing for a user to have
   * authorized per repo before a session will start.
   */
  prepareGitHub(repos: string[]): Promise<GitHubSetup> {
    return tokenGitHubSetup(repos);
  }

  // MARK: - Auto-naming

  generateRenameToken(): Promise<string> {
    return this.service.generateRenameToken();
  }

  // MARK: - Naming and reachability

  destinationForVMName(name: string): string {
    return `${name}.exe.xyz`;
  }

  webURLForDestination(destination: string): string | null {
    return `https://${destination}`;
  }

  parseReflectedName(output: string): string | null {
    try {
      const parsed = JSON.parse(output) as unknown;
      if (typeof parsed !== "object" || parsed === null) return null;
      const name = (parsed as Record<string, unknown>).name;
      return typeof name === "string" && name ? name : null;
    } catch {
      return null;
    }
  }

  // MARK: - Transport

  transportFor(destination: string): RemoteTransport {
    return new SSHTransport(destination);
  }

  hostEnvironmentSetup(_environment: EnvVar[]): string {
    // exe.dev sets host env via `new --env`, so the bootstrap injects nothing.
    return "";
  }
}

/**
 * An `ExeVM` as the app-agnostic `RemoteVMRecord`. `ssh_dest` is the transport
 * handle; the web URL is the SSH host over HTTPS.
 */
export function recordFromExeVM(vm: ExeVM): RemoteVMRecord {
  const name = vm.vm_name ?? vm.ssh_dest ?? "unknown";
  const destination = vm.ssh_dest ?? (vm.vm_name ? `${vm.vm_name}.exe.xyz` : "unknown");
  return {
    name,
    destination,
    webURL: `https://${destination}`,
    status: vm.status ?? null,
    provider: "exe",
  };
}
